test/tests/functional/pbs_cgroups_hook.py

```diff
@@ -0,0 +1,1083 @@
+        {
+            "enabled"         : true,
+            "exclude_hosts"   : [],
+            "exclude_vntypes" : ["no_cgroups_cpus"]
+        },
+        "devices":
+        {
+            "enabled"         : false
+        },
+        "hugetlb":
+        {
+            "enabled"         : false
+        },
+        "memory":
+        {
+            "enabled"         : true,
+            "default"         : "256MB",
+            "reserve_amount"  : "100MB",
+            "exclude_hosts"   : [],
+            "exclude_vntypes" : ["no_cgroups_mem"]
+        },
+        "memsw":
+        {
+            "enabled"         : true,
+            "default"         : "256MB",
+            "reserve_amount"  : "90MB",
+            "exclude_hosts"   : [],
+            "exclude_vntypes" : []
+        }
+    }
+}
+"""
+        Job.dflt_attributes[ATTR_k] = 'oe'
+        # Increase the log level
+        a = {'log_events': '2047'}
+        self.server.manager(MGR_CMD_SET, SERVER, a, expect=True)
+        # Configure the scheduler to schedule using vmem
+        a = {'resources': 'ncpus,mem,vmem,host,vnode'}
+        self.scheduler.set_sched_config(a)
+        # Configure the mom
+        self.momA = self.moms.values()[0]
+        c = {'$logevent': '0xffffffff', '$clienthost': self.server.name,
+             '$min_check_poll': 8, '$max_check_poll': 12}
+        self.momA.add_config(c)
+        # Create resource as root
+        attr = {'type': 'long', 'flag': 'nh'}
+        rc = self.server.manager(MGR_CMD_CREATE, RSC, attr, id='nmics',
+                                 runas='root', logerr=False)
+        rc = self.server.manager(MGR_CMD_CREATE, RSC, attr, id='ngpus',
+                                 runas='root', logerr=False)
+        # Import the hook
+        self.hook_name = 'pbs_cgroups'
+        hook_url = 'https://raw.githubusercontent.com/' + \
+            'PBSPro/pbspro/master/src/hooks/cgroups/pbs_cgroups.py'
+        self.load_hook(hook_url)
+        # Enable the cgroups hook
+        conf = {'enabled': 'True', 'freq': 2}
+        self.server.manager(MGR_CMD_SET, HOOK, conf, self.hook_name)
+        # Restart mom so exechost_startup hook is run
+        time.sleep(2)
+        self.mom.signal('-HUP')
+
+    def get_paths(self):
+        paths = {}
+        # Loop through the mounts and collect the ones for cgroups
+        with open(os.path.join(os.sep, "proc", "mounts"), 'r') as fd:
+            for line in fd:
+                entries = line.split()
+                if entries[2] != "cgroup":
+                    continue
+                flags = entries[3].split(',')
+                subsys = os.path.basename(entries[1])
+                paths[subsys] = entries[1]
+                if 'memory' in flags:
+                    paths['memsw'] = paths['memory']
+        return paths
+
+    def random_name(self, length=16):
+        return str().join(random.choice(string.ascii_lowercase)
+                          for _ in range(length))
```

Comment:
The syntax being used is called list comprehension. It is covered in the Python guide. A performance study is located here: [https://waymoot.org/home/python_string/]